content/docs/rex_book/writing_modules/writing_custom_resources.md

```diff
@@ -52,184 +52,192 @@ it between projects.
-  };
-
-  # get the right provider if it is not defined via the operating system
-  # and the list from above.
-  my $provider =
-    param_lookup( "provider", case ( lc(operating_system), $__provider ) );
-
-  # load the provider class
-  $provider->require;
-  
-  # create a new instance of the provider class
-  my $provider_o = $provider->new();
-
-  # and execute the requested state.
-  if ( $rule_config->{ensure} eq "present" ) {
-    if ( $provider_o->present($rule_config) ) {
-      emit created, "HelloWorld::greet resource created.";
-    }
-  }
-  elsif ( $rule_config->{ensure} eq "absent" ) {
-    if ( $provider_o->absent($rule_config) ) {
-      emit removed, "HelloWorld::greet resource removed.";
-    }
-  }
-  else {
-    die "Error: $rule_config->{ensure} not a valid option for 'ensure'.";
-  }
-
-};
-
-1; # this need to be the last line in the file</code></pre>
+    ```perl
+    package HelloWorld;
+    
+    use strict;
+    use warnings;
+    
+    use Rex -minimal; # for Rex < 1.4 use Rex -base;
+    use Rex::Resource::Common; # load resource functions
+    
+    # load the Gather functions, so we have the `operating_system` 
+    # function.
+    use Rex::Commands::Gather;
+    
+    use Carp;
+    
+    # list the available providers
+    my $__provider = {
+      default => "HelloWorld::greet::Provider::default",
+      CentOS  => "HelloWorld::greet::Provider::centos",
+    };
+    
+    # create a resource HelloWorld::greet
+    resource "greet", sub {
+      my $rule_name = resource_name;
+    
+      my $rule_config = {
+        ensure  => param_lookup( "ensure", "present" ),
+        message => param_lookup( "message", "<default value>" ),
+      };
+    
+      # get the right provider if it is not defined via the operating system
+      # and the list from above.
+      my $provider =
+        param_lookup( "provider", case ( lc(operating_system), $__provider ) );
+    
+      # load the provider class
+      $provider->require;
+      
+      # create a new instance of the provider class
+      my $provider_o = $provider->new();
+    
+      # and execute the requested state.
+      if ( $rule_config->{ensure} eq "present" ) {
+        if ( $provider_o->present($rule_config) ) {
+          emit created, "HelloWorld::greet resource created.";
+        }
+      }
+      elsif ( $rule_config->{ensure} eq "absent" ) {
+        if ( $provider_o->absent($rule_config) ) {
+          emit removed, "HelloWorld::greet resource removed.";
+        }
+      }
+      else {
+        die "Error: $rule_config->{ensure} not a valid option for 'ensure'.";
+      }
+    
+    };
+    
+    1; # this need to be the last line in the file
+    ```
 
 Now, after creating the resource definition we need to create our providers.
 To do so, you need to create the following directory.
 
-<pre><code class="bash">$ mkdir -p lib/HelloWorld/greet/Provider</code></pre>
+    $ mkdir -p lib/HelloWorld/greet/Provider
```

Comment:
shouldn't this be enclosed with a ` ```bash` block?

Reply:
Technically, yes, it's _possible_ to enclose the bash snippets with
~~~
```bash
[...code...]
```
~~~

but I decided against it for now because:
1. bash highlighting for CLI commands rarely add any actual highlighting, they seem to be more useful for scripts
2. they also add distracting highlighting in the command outputs, when the example containes them too
3. adds two extra lines to the bash examples that are mostly short (1-2 lines, so triples/doubles code block length for no value)

So I've left them as-is with the default indent-based style.

ps.: in 2 or 3 occasions I've left out the YAML highlighting as well, for the same reasons _and_ because `Syntax::Highlight::Engine::Kate` doesn't seem to support it.

Reply:
Ok. Then I think this is ready for merging.